### ISSUE TITLE
Proposal: Standardize and disambiguate ONNX output tensor names across tasks and export modes

### ISSUE THREAD

**Issue by levipereira** [ultralytics/ultralytics#24265]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Currently, the ONNX exporter in ultralytics/utils/export/engine.py uses the generic name output0 (and output1 for segmentation masks) as the default output tensor name for all tasks (detect, segment, pose, obb, classify) and both export modes (raw head output and end-to-end with NMS)

The name `output0` carries no semantic information about:

What task the tensor belongs to (detection logits vs. classification probabilities vs. pose keypoints)
Which export mode produced it (raw concatenated head output [B, 4+nc, N] vs. end-to-end post-NMS detections)
What layout the consumer should expect

This makes downstream tooling (TensorRT builders, DeepStream parsers, Triton config generators, custom C++ inference wrappers) unable to auto-detect the model contract by inspecting tensor names. Integrators end up hardcoding assumptions or relying on shape heuristics, which break silently when the same pipeline receives a model exported with a different flag combination.

Proposal
Adopt a semantic, stable naming scheme for ONNX output tensors that encodes the export contract. 
Example scheme (open to discussion):
Raw head output (nms=False):

`out_raw` — single tensor with concatenated head output [B, 4+nc(+extras), N]
`out_proto` — segmentation mask prototypes (replaces output1 for segment)

End-to-end (nms=True / end2end=True):
`out_e2e` — final detections tensor [B, max_det, 6]

Either way, the key property is: the output name alone should identify the export mode.

A DeepStream NvDsInferParseCustomFunc implementation today has to branch on shape and metadata to figure out which schema it is looking at, even though all of them arrive with the same tensor name output0. A semantic name would let a single parser dispatch deterministically on the output name alone, which is cheaper, safer, and reusable across models.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @levipereira, thank you for your thoughtful feature request and for searching existing issues first 🚀 This is an automated response to help get things moving quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this is a feature request 💡, it would be helpful if you could share any additional implementation details, expected naming behavior across tasks/export modes, and example downstream integrations or parser expectations that would benefit from the proposed ONNX output naming standardization.

If this is or becomes a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
This looks reasonable; today ONNX export still defaults to generic `output0`/`output1`, while the actual contract changes between raw and end-to-end exports as shown in the [export docs](https://docs.ultralytics.com/modes/export/) and the current `torch2onnx()` default names in the [engine reference](https://docs.ultralytics.com/reference/utils/export/engine/), so we can treat this as a valid enhancement request for clearer task/mode-specific tensor names with backward compatibility in mind.
